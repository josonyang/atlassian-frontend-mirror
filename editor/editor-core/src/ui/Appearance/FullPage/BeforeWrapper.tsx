/** @jsx jsx */
import type { ReactElement } from 'react';

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
import { css, jsx } from '@emotion/react';

import { token } from '@atlaskit/tokens';

const beforePrimaryToolbarPluginWrapperStyles = css({
	display: 'flex',
	marginRight: token('space.100', '8px'),
	flexGrow: 1,
	justifyContent: 'flex-end',
	alignItems: 'center',
});

type ReactComponents = ReactElement<any> | ReactElement<any>[];

// Duplicate of the wrapper from `editor-plugins/before-primary-toolbar` used
// only in `FullPageToolbar` to decouple the plugin from the main toolbar
export const BeforePrimaryToolbarWrapper = (props: {
	beforePrimaryToolbarComponents: ReactComponents | undefined;
}) => (
	<div
		css={beforePrimaryToolbarPluginWrapperStyles}
		data-testid={'before-primary-toolbar-components-plugin'}
	>
		{props.beforePrimaryToolbarComponents}
	</div>
);
