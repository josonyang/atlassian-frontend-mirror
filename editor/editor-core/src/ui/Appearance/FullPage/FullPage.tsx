/**
 * @jsxRuntime classic
 * @jsx jsx
 */
import React, { useEffect, useMemo, useRef, useState } from 'react';

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
import { jsx } from '@emotion/react';
import { IntlProvider } from 'react-intl-next';

import { browser } from '@atlaskit/editor-common/browser';
import { useSharedPluginStateWithSelector } from '@atlaskit/editor-common/hooks';
import type { OptionalPlugin } from '@atlaskit/editor-common/types';
import { ContextPanelWidthProvider } from '@atlaskit/editor-common/ui';
import { useSharedPluginStateSelector } from '@atlaskit/editor-common/use-shared-plugin-state-selector';
import type { EditorViewModePlugin } from '@atlaskit/editor-plugins/editor-viewmode';
import type { InteractionPlugin } from '@atlaskit/editor-plugins/interaction';
import type { PrimaryToolbarPlugin } from '@atlaskit/editor-plugins/primary-toolbar';
import type { SelectionToolbarPlugin } from '@atlaskit/editor-plugins/selection-toolbar';
import type { ToolbarPlugin } from '@atlaskit/editor-plugins/toolbar';
import { FULL_PAGE_EDITOR_TOOLBAR_HEIGHT } from '@atlaskit/editor-shared-styles';
import { fg } from '@atlaskit/platform-feature-flags';
import { editorExperiment } from '@atlaskit/tmp-editor-statsig/experiments';

import type { EditorAppearanceComponentProps, PrimaryToolbarComponents } from '../../../types';
import { getPrimaryToolbarComponents } from '../../Toolbar/getPrimaryToolbarComponents';

import { FullPageContentArea } from './FullPageContentArea';
import type { ToolbarEditorPlugins } from './FullPageToolbar';
import { FullPageToolbar } from './FullPageToolbar';
import { FullPageToolbarNext } from './FullPageToolbarNext';
import { fullPageEditorWrapper } from './StyledComponents';
import { type ScrollContainerRefs } from './types';

const useShowKeyline = (contentAreaRef: React.MutableRefObject<ScrollContainerRefs | null>) => {
	const [showKeyline, setShowKeyline] = useState<boolean>(false);

	useEffect(() => {
		if (!contentAreaRef.current?.contentArea) {
			return;
		}

		const intersection = new IntersectionObserver(
			([entry]) => {
				setShowKeyline(
					!entry.isIntersecting && entry.boundingClientRect.top < entry.intersectionRect.top,
				);
			},
			{
				root: undefined,
				// Safari seems to miss events (on fast scroll) sometimes due
				// to differences in IntersectionObserver behaviour between browsers.
				// By lowering the threshold a little it gives Safari more
				// time to catch these events.
				threshold: browser.safari ? 0.98 : 1,
			},
		);

		intersection.observe(contentAreaRef.current.contentArea);

		return () => {
			intersection.disconnect();
		};
	}, [contentAreaRef]);

	return showKeyline;
};

export type ComponentProps = EditorAppearanceComponentProps<
	[
		OptionalPlugin<EditorViewModePlugin>,
		OptionalPlugin<PrimaryToolbarPlugin>,
		OptionalPlugin<SelectionToolbarPlugin>,
		OptionalPlugin<InteractionPlugin>,
		OptionalPlugin<ToolbarPlugin>,
		...ToolbarEditorPlugins,
	]
>;

const hasCustomComponents = (components?: PrimaryToolbarComponents) => {
	if (!components) {
		return false;
	}

	if ('before' in components) {
		return (
			(Array.isArray(components.before) && components.before.length > 0) ||
			!!components.before ||
			(Array.isArray(components.after) && components.after.length > 0) ||
			!!components.after
		);
	}

	return true;
};

export const FullPageEditor = (props: ComponentProps) => {
	const wrapperElementRef = useMemo(() => props.innerRef, [props.innerRef]);
	const scrollContentContainerRef = useRef<ScrollContainerRefs | null>(null);
	const showKeyline = useShowKeyline(scrollContentContainerRef);
	const editorAPI = props.editorAPI;

	const state = useSharedPluginStateWithSelector(
		editorAPI,
		['primaryToolbar', 'interaction', 'editorViewMode'],
		(states) => ({
			primaryToolbarComponents: states.primaryToolbarState?.components,
			interactionState: states.interactionState?.interactionState,
			editorViewMode: states.editorViewModeState?.mode,
		}),
	);

	const interactionState = state.interactionState;

	const primaryToolbarState = getPrimaryToolbarComponents(
		editorAPI,
		state.primaryToolbarComponents,
	);

	const hasHadInteraction = interactionState !== 'hasNotHadInteraction';

	let toolbarDocking = useSharedPluginStateSelector(editorAPI, 'selectionToolbar.toolbarDocking', {
		disabled: fg('platform_editor_use_preferences_plugin'),
	});
	if (!fg('platform_editor_use_preferences_plugin')) {
		if (!toolbarDocking) {
			// This is a workaround for the rendering issue with the selection toolbar
			// where using useSharedPluginStateSelector or useSharedPluginState the state are not
			// available when the editor is first loaded. and cause the toolbar to blink.
			const defaultDocking = props.__livePage ? 'none' : 'top';
			toolbarDocking =
				editorAPI?.selectionToolbar?.sharedState.currentState()?.toolbarDocking ?? defaultDocking;
		}
	}

	let { toolbarDockingPosition } =
		useSharedPluginStateSelector(editorAPI, 'userPreferences.preferences', {
			disabled: !fg('platform_editor_use_preferences_plugin'),
		}) || {};
	if (fg('platform_editor_use_preferences_plugin')) {
		if (!toolbarDockingPosition) {
			// This is a workaround for the rendering issue with the selection toolbar
			// when using useSharedPluginStateWithSelector the state is not yet
			// available when the editor is first loaded.
			// This causes the toolbar to blink creating a layout shift.
			const defaultDockingPosition = props.__livePage ? 'none' : 'top';
			toolbarDockingPosition =
				editorAPI?.userPreferences?.actions.getUserPreferences()?.toolbarDockingPosition ??
				defaultDockingPosition;
		}
	}

	let primaryToolbarComponents = props.primaryToolbarComponents;

	if (Array.isArray(primaryToolbarState?.components) && Array.isArray(primaryToolbarComponents)) {
		primaryToolbarComponents = primaryToolbarState.components.concat(primaryToolbarComponents);
	}

	let isEditorToolbarHidden = state.editorViewMode === 'view';

	const { customPrimaryToolbarComponents } = props;

	if (editorExperiment('platform_editor_controls', 'variant1', { exposure: true })) {
		if (fg('platform_editor_use_preferences_plugin')) {
			// need to check if the toolbarDockingPosition is set to 'none' or 'top'
			if (toolbarDockingPosition === 'none') {
				primaryToolbarComponents = [];

				if (!hasCustomComponents(customPrimaryToolbarComponents)) {
					isEditorToolbarHidden = true;
				}
			}
		} else {
			if (toolbarDocking === 'none') {
				primaryToolbarComponents = [];

				if (!hasCustomComponents(customPrimaryToolbarComponents)) {
					isEditorToolbarHidden = true;
				}
			}
		}
	}

	const popupsBoundariesElement =
		props.popupsBoundariesElement || scrollContentContainerRef?.current?.containerArea || undefined;

	return (
		<ContextPanelWidthProvider>
			<div
				// eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/design-system/consistent-css-prop-usage -- Ignored via go/DSP-18766
				css={fullPageEditorWrapper}
				// eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- Ignored via go/DSP-18766
				className="akEditor"
				ref={wrapperElementRef}
				style={
					{
						// eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values
						'--ak-editor-fullpage-toolbar-height': FULL_PAGE_EDITOR_TOOLBAR_HEIGHT(),
					} as React.CSSProperties
				}
			>
				{!isEditorToolbarHidden &&
					(editorExperiment('platform_editor_toolbar_aifc', true) ? (
						<IntlProvider locale="en">
							<FullPageToolbarNext
								toolbarDockingPosition={toolbarDockingPosition ?? toolbarDocking}
								beforeIcon={props.primaryToolbarIconBefore}
								editorAPI={editorAPI}
								editorView={props.editorView}
								popupsMountPoint={props.popupsMountPoint}
								popupsBoundariesElement={props.popupsBoundariesElement}
								popupsScrollableElement={props.popupsScrollableElement}
								showKeyline={showKeyline}
								customPrimaryToolbarComponents={props.customPrimaryToolbarComponents}
							/>
						</IntlProvider>
					) : (
						<FullPageToolbar
							appearance={props.appearance}
							editorAPI={editorAPI}
							beforeIcon={props.primaryToolbarIconBefore}
							collabEdit={props.collabEdit}
							containerElement={scrollContentContainerRef.current?.scrollContainer ?? null}
							customPrimaryToolbarComponents={props.customPrimaryToolbarComponents}
							disabled={!!props.disabled}
							dispatchAnalyticsEvent={props.dispatchAnalyticsEvent}
							editorActions={props.editorActions}
							editorDOMElement={props.editorDOMElement}
							// Ignored via go/ees005
							// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
							editorView={props.editorView!}
							// Ignored via go/ees005
							// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
							eventDispatcher={props.eventDispatcher!}
							hasMinWidth={props.enableToolbarMinWidth}
							popupsBoundariesElement={props.popupsBoundariesElement}
							popupsMountPoint={props.popupsMountPoint}
							popupsScrollableElement={props.popupsScrollableElement}
							primaryToolbarComponents={primaryToolbarComponents}
							providerFactory={props.providerFactory}
							showKeyline={showKeyline}
							featureFlags={props.featureFlags}
						/>
					))}
				<FullPageContentArea
					editorAPI={editorAPI}
					ref={scrollContentContainerRef}
					appearance={props.appearance}
					contentComponents={props.contentComponents}
					contextPanel={props.contextPanel}
					customContentComponents={props.customContentComponents}
					disabled={props.disabled}
					dispatchAnalyticsEvent={props.dispatchAnalyticsEvent}
					editorActions={props.editorActions}
					editorDOMElement={props.editorDOMElement}
					// Ignored via go/ees005
					// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
					editorView={props.editorView!}
					eventDispatcher={props.eventDispatcher}
					popupsBoundariesElement={popupsBoundariesElement}
					popupsMountPoint={props.popupsMountPoint}
					popupsScrollableElement={props.popupsScrollableElement}
					providerFactory={props.providerFactory}
					wrapperElement={wrapperElementRef?.current ?? null}
					pluginHooks={props.pluginHooks}
					featureFlags={props.featureFlags}
					isEditorToolbarHidden={isEditorToolbarHidden}
					viewMode={state.editorViewMode}
					hasHadInteraction={hasHadInteraction}
				/>
			</div>
		</ContextPanelWidthProvider>
	);
};
