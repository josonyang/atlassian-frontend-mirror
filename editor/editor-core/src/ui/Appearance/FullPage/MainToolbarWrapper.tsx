/**
 * @jsxRuntime classic
 * @jsx jsx
 */
import React from 'react';

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
import { css, jsx } from '@emotion/react';

import {
	akEditorFloatingDialogZIndex,
	akEditorSwoopCubicBezier,
} from '@atlaskit/editor-shared-styles';
import { fg } from '@atlaskit/platform-feature-flags';
import { componentWithCondition } from '@atlaskit/platform-feature-flags-react';
import { expValEquals } from '@atlaskit/tmp-editor-statsig/exp-val-equals';
import { token } from '@atlaskit/tokens';

import {
	mainToolbarStyle as mainToolbarStyleDynamic,
	MAXIMUM_TWO_LINE_TOOLBAR_BREAKPOINT,
} from './MainToolbar';

// Base styles that don't depend on feature flags
const baseToolbarStyles = css({
	position: 'relative',
	alignItems: 'center',
	boxShadow: 'none',
	borderBottom: `1px solid ${token('color.border')}`,
	// eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values
	transition: `box-shadow 200ms ${akEditorSwoopCubicBezier}`,
	// eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values, @atlaskit/ui-styling-standard/no-imported-style-values
	zIndex: akEditorFloatingDialogZIndex,
	// eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values, @atlaskit/ui-styling-standard/no-imported-style-values
	display: 'flex',
	height: 'var(--ak-editor-fullpage-toolbar-height)',
	flexShrink: 0,
	backgroundColor: token('elevation.surface'),
	// eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors
	'& object': {
		// eslint-disable-next-line @atlaskit/ui-styling-standard/no-important-styles
		height: '0 !important',
	},
});

const flexibleIconSize = css({
	// eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors
	'& span svg': {
		maxWidth: '100%',
	},
});

interface MainToolbarWrapperProps {
	children: React.ReactNode;
	'data-testid'?: string;
	showKeyline: boolean;
	twoLineEditorToolbar: boolean;
}

// box-shadow is overriden by the mainToolbar
const mainToolbarWithKeyline = css({
	// eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
	boxShadow: `${token('elevation.shadow.overflow')}`,
});

const mainToolbarTwoLineStyle = css({
	// eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values, @atlaskit/ui-styling-standard/no-imported-style-values -- Ignored via go/DSP-18766
	[`@media (max-width: ${MAXIMUM_TWO_LINE_TOOLBAR_BREAKPOINT}px)`]: {
		flexWrap: 'wrap',
		// eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/ui-styling-standard/no-unsafe-values -- Ignored via go/DSP-18766
		height: `calc(var(--ak-editor-fullpage-toolbar-height) * 2)`,
	},
});

const MainToolbarWrapperNext = ({
	showKeyline,
	twoLineEditorToolbar,
	children,
	'data-testid': testId,
}: MainToolbarWrapperProps) => {
	return (
		<div
			// eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage
			css={[
				baseToolbarStyles,
				fg('platform-visual-refresh-icons') && flexibleIconSize,
				showKeyline && mainToolbarWithKeyline,
				twoLineEditorToolbar && mainToolbarTwoLineStyle,
			]}
			data-testid={testId}
		>
			{children}
		</div>
	);
};

/**
 * Original version of the toolbar wrapper using dynamic styles
 */
const MainToolbarWrapperOld = ({
	showKeyline,
	twoLineEditorToolbar,
	children,
	'data-testid': testId,
}: MainToolbarWrapperProps) => {
	return (
		<div
			css={
				// eslint-disable-next-line @atlaskit/design-system/consistent-css-prop-usage, @atlaskit/ui-styling-standard/no-imported-style-values
				mainToolbarStyleDynamic(showKeyline, twoLineEditorToolbar)
			}
			data-testid={testId}
		>
			{children}
		</div>
	);
};

/**
 * Wrapper component for the main toolbar that handles feature flag based styling
 * @example
 * <MainToolbarWrapper showKeyline={true} twoLineEditorToolbar={false}>
 *   <ToolbarContent />
 * </MainToolbarWrapper>
 */
export const MainToolbarWrapper = componentWithCondition(
	() => expValEquals('platform_editor_core_static_emotion_non_central', 'isEnabled', true),
	MainToolbarWrapperNext,
	MainToolbarWrapperOld,
);
