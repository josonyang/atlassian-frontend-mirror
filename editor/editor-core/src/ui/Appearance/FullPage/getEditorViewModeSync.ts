import type { ViewMode, EditorViewModePluginState } from '@atlaskit/editor-plugins/editor-viewmode';
import { fg } from '@atlaskit/platform-feature-flags';

import type { ComponentProps } from './FullPage';

// Retrieve the initial config value from the editor view plugin preset
const getViewModeSync = (preset: ComponentProps['preset']) => {
	if (!fg('platform_editor_sync_editor_view_mode_state')) {
		return undefined;
	}
	// @ts-expect-error - data is a private property
	const editorViewModePlugin = preset.data.find((p) => {
		if (Array.isArray(p)) {
			const [plugin, config] = p;
			return plugin({ config }).name === 'editorViewMode';
		} else {
			return p?.({ config: undefined })?.name === 'editorViewMode';
		}
	});
	// If we have set a default value on the view mode (index 1) then we can return the mode, otherwise we haven't set by default
	return Array.isArray(editorViewModePlugin) ? editorViewModePlugin[1]?.mode : undefined;
};

// Temporary to ensure view mode is always set synchronously
// TODO: ED-27194 - Make editorAPI be set synchronously and we can remove this code
export const getEditorViewMode = (
	editorViewModeState: EditorViewModePluginState | undefined | null,
	preset: ComponentProps['preset'],
): ViewMode => {
	if (!editorViewModeState?.mode) {
		return getViewModeSync(preset);
	}
	return editorViewModeState?.mode;
};
