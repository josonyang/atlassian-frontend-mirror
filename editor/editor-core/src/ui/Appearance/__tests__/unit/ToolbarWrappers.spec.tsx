import React from 'react';

import { render } from '@testing-library/react';

import { setupEditorExperiments } from '@atlaskit/tmp-editor-statsig/setup';

import { BeforePrimaryToolbarWrapper } from '../../FullPage/BeforeWrapper';
import {
	MainToolbarForFirstChildWrapper as FirstChildWrapper,
	MainToolbarForSecondChildWrapper as SecondChildWrapper,
} from '../../FullPage/CustomToolbarWrapper';
import { MainToolbarWrapper } from '../../FullPage/MainToolbarWrapper';

describe('Toolbar wrapper components', () => {
	describe('MainToolbarWrapper', () => {
		const renderWrapper = (twoLine = false, keyline = false) =>
			render(
				<MainToolbarWrapper showKeyline={keyline} twoLineEditorToolbar={twoLine}>
					<div>child</div>
				</MainToolbarWrapper>,
			);

		afterEach(() => {
			// reset overrides
			setupEditorExperiments('test', {});
		});

		it.each([false, true])('should match snapshot when static emotion = %p', (staticEmotion) => {
			setupEditorExperiments('test', {
				// toggle experiment variant
				platform_editor_core_static_emotion_non_central: staticEmotion,
			});

			const { container } = renderWrapper();
			expect(container).toMatchSnapshot();
		});

		it('should match snapshot – two-line toolbar', () => {
			const { container } = renderWrapper(true);
			expect(container).toMatchSnapshot();
		});

		it('should match snapshot – keyline visible', () => {
			const { container } = renderWrapper(false, true);
			expect(container).toMatchSnapshot();
		});
	});

	describe('MainToolbar child wrappers', () => {
		it.each([false, true])('FirstChildWrapper snapshot twoLine=%p', (twoLine) => {
			const { container } = render(
				<FirstChildWrapper twoLineEditorToolbar={twoLine}>
					<div>child</div>
				</FirstChildWrapper>,
			);
			expect(container).toMatchSnapshot();
		});

		it.each([false, true])('SecondChildWrapper snapshot twoLine=%p', (twoLine) => {
			const { container } = render(
				<SecondChildWrapper twoLineEditorToolbar={twoLine}>
					<div>child</div>
				</SecondChildWrapper>,
			);
			expect(container).toMatchSnapshot();
		});
	});

	describe('BeforePrimaryToolbarWrapper', () => {
		const renderWrapper = (component?: React.ReactElement) =>
			render(<BeforePrimaryToolbarWrapper beforePrimaryToolbarComponents={component} />);

		afterEach(() => {
			// reset overrides
			setupEditorExperiments('test', {});
		});

		it('should match snapshot – with components', () => {
			const { container } = renderWrapper(<div>before</div>);
			expect(container).toMatchSnapshot();
		});

		it('should match snapshot – without components', () => {
			const { container } = renderWrapper(undefined);
			expect(container).toMatchSnapshot();
		});
	});
});
