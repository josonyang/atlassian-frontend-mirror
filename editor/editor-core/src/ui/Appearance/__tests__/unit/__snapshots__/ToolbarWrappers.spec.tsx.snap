// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolbar wrapper components BeforePrimaryToolbarWrapper should match snapshot – with components 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: flex-end;
  justify-content: flex-end;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div>
  <div
    class="emotion-0"
    data-testid="before-primary-toolbar-components-plugin"
  >
    <div>
      before
    </div>
  </div>
</div>
`;

exports[`Toolbar wrapper components BeforePrimaryToolbarWrapper should match snapshot – without components 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: flex-end;
  justify-content: flex-end;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div>
  <div
    class="emotion-0"
    data-testid="before-primary-toolbar-components-plugin"
  />
</div>
`;

exports[`Toolbar wrapper components MainToolbar child wrappers FirstChildWrapper snapshot twoLine=false 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

@media (max-width: 0px) {
  .emotion-0 {
    grid-column: 1;
    grid-row: 1;
  }
}

<div>
  <div
    class="emotion-0"
  >
    <div>
      child
    </div>
  </div>
</div>
`;

exports[`Toolbar wrapper components MainToolbar child wrappers FirstChildWrapper snapshot twoLine=true 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

@media (max-width: 0px) {
  .emotion-0 {
    grid-column: 1;
    grid-row: 1;
  }
}

@media (max-width: 868px) {
  .emotion-0 {
    -webkit-flex: 1 1 100%;
    -ms-flex: 1 1 100%;
    flex: 1 1 100%;
    height: var(--ds-space-500, 40px);
    -webkit-box-pack: end;
    -ms-flex-pack: end;
    -webkit-justify-content: flex-end;
    justify-content: flex-end;
    min-width: -webkit-fit-content;
    min-width: -moz-fit-content;
    min-width: fit-content;
  }
}

<div>
  <div
    class="emotion-0"
  >
    <div>
      child
    </div>
  </div>
</div>
`;

exports[`Toolbar wrapper components MainToolbar child wrappers SecondChildWrapper snapshot twoLine=false 1`] = `
.emotion-0 {
  min-width: -webkit-fit-content;
  min-width: -moz-fit-content;
  min-width: fit-content;
}

<div>
  <div
    class="emotion-0"
  >
    <div>
      child
    </div>
  </div>
</div>
`;

exports[`Toolbar wrapper components MainToolbar child wrappers SecondChildWrapper snapshot twoLine=true 1`] = `
.emotion-0 {
  min-width: -webkit-fit-content;
  min-width: -moz-fit-content;
  min-width: fit-content;
}

@media (max-width: 868px) {
  .emotion-0 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-flex: 1;
    -webkit-flex-grow: 1;
    -ms-flex-positive: 1;
    flex-grow: 1;
    -webkit-flex: 1 1 100%;
    -ms-flex: 1 1 100%;
    flex: 1 1 100%;
    margin: auto;
    height: var(--ds-space-500, 40px);
    min-width: 0;
  }
}

<div>
  <div
    class="emotion-0"
  >
    <div>
      child
    </div>
  </div>
</div>
`;

exports[`Toolbar wrapper components MainToolbarWrapper should match snapshot when static emotion = false 1`] = `
.emotion-0 {
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-shadow: none;
  border-bottom: 1px solid var(--ds-border, #091E4224);
  -webkit-transition: box-shadow 200ms cubic-bezier(0.15, 1, 0.3, 1);
  transition: box-shadow 200ms cubic-bezier(0.15, 1, 0.3, 1);
  z-index: 510;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: var(--ds-space-500, 40px);
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  background-color: var(--ds-surface, white);
}

.emotion-0 object {
  height: 0!important;
}

@media (max-width: 0px) {
  .emotion-0 {
    display: grid;
    height: calc(var(--ds-space-500, 40px) * 2);
  }
}

<div>
  <div
    class="emotion-0"
  >
    <div>
      child
    </div>
  </div>
</div>
`;

exports[`Toolbar wrapper components MainToolbarWrapper should match snapshot when static emotion = true 1`] = `
.emotion-0 {
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-shadow: none;
  border-bottom: 1px solid var(--ds-border, #091E4224);
  -webkit-transition: box-shadow 200ms cubic-bezier(0.15, 1, 0.3, 1);
  transition: box-shadow 200ms cubic-bezier(0.15, 1, 0.3, 1);
  z-index: 510;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: var(--ak-editor-fullpage-toolbar-height);
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  background-color: var(--ds-surface, #FFFFFF);
}

.emotion-0 object {
  height: 0!important;
}

<div>
  <div
    class="emotion-0"
  >
    <div>
      child
    </div>
  </div>
</div>
`;

exports[`Toolbar wrapper components MainToolbarWrapper should match snapshot – keyline visible 1`] = `
.emotion-0 {
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-shadow: none;
  border-bottom: 1px solid var(--ds-border, #091E4224);
  -webkit-transition: box-shadow 200ms cubic-bezier(0.15, 1, 0.3, 1);
  transition: box-shadow 200ms cubic-bezier(0.15, 1, 0.3, 1);
  z-index: 510;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: var(--ds-space-500, 40px);
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  background-color: var(--ds-surface, white);
  box-shadow: var(--ds-shadow-overflow, 0px 0px 8px #091E4228, 0px 0px 1px #091E421e);
}

.emotion-0 object {
  height: 0!important;
}

@media (max-width: 0px) {
  .emotion-0 {
    display: grid;
    height: calc(var(--ds-space-500, 40px) * 2);
  }
}

<div>
  <div
    class="emotion-0"
  >
    <div>
      child
    </div>
  </div>
</div>
`;

exports[`Toolbar wrapper components MainToolbarWrapper should match snapshot – two-line toolbar 1`] = `
.emotion-0 {
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-shadow: none;
  border-bottom: 1px solid var(--ds-border, #091E4224);
  -webkit-transition: box-shadow 200ms cubic-bezier(0.15, 1, 0.3, 1);
  transition: box-shadow 200ms cubic-bezier(0.15, 1, 0.3, 1);
  z-index: 510;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: var(--ds-space-500, 40px);
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  background-color: var(--ds-surface, white);
}

.emotion-0 object {
  height: 0!important;
}

@media (max-width: 0px) {
  .emotion-0 {
    display: grid;
    height: calc(var(--ds-space-500, 40px) * 2);
  }
}

@media (max-width: 868px) {
  .emotion-0 {
    -webkit-box-flex-wrap: wrap;
    -webkit-flex-wrap: wrap;
    -ms-flex-wrap: wrap;
    flex-wrap: wrap;
    height: calc(var(--ds-space-500, 40px) * 2);
  }
}

<div>
  <div
    class="emotion-0"
  >
    <div>
      child
    </div>
  </div>
</div>
`;
